Fix issues with `Win32` / `x86` compilation

maxgolov opened the pull request:
Fixes #842

NOTE: I have only fixed and tested the essential common core parts. There could be other fixes necessary for `Win32` in the optional parts that are not planned for inclusion in `v1.0` GA. These should be covered separately by setting up a build loop for `Win32`, which we do not have in CI. We only test `x64` at the moment.

## Changes

- code fix for proper type cast.
- detect arch and invoke the installation of missing deps for the proper architecture (by default it was installing deps for `x64-windows` , but if CMake is targeting `Win32` - we need to install for `x86-windows`).
- gRPC example uses some funky target that is not even applicable; looks like a bug, and I'm not sure why it slipped thru the cracks before. Found it while testing x86 build.

maxgolov commented on CMakeLists.txt:
```diff
@@ -112,7 +127,9 @@ if(WIN32)
     if(MSVC)
-      # GTest bug: https://github.com/google/googletest/issues/860
+      # Uncomment for strict Level4 warnings if building for test:
+      # set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W4") GTest bug:
```
I hate our auto-formatter, lol..

ThomsonTan commented on CMakeLists.txt:
```diff
@@ -9,6 +9,21 @@ project(opentelemetry-cpp)
+  if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+    # 64 bits
+    set(ARCH x64)
+  elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
```
Also check target is Intel?

maxgolov replied:
Yes, I agree we should.. My point was mostly that if someone builds for non-Intel (arm, risc, ppc), they'd have to specify the `ARCH` variable manually. Since this variable is currently only being used for Windows dependencies and/or with vcpkg, possible supported architectures are `[x86, x64, arm, arm64]`. I can try to improve this to handle Intel and ARM, but with a disclaimer above that any other non-Intel and non-ARM architectures should go thru `if(DEFINED ENV{ARCH})` path above. Let me think it over.

maxgolov replied:
@ThomsonTan - I carefully reviewed the options. The challenge is that on Windows - CMake does not necessarily properly populate the target architecture. This is described in CMake documentation. Best we can come up with:

- assume that users know what arch they are compiling. And if this is ARM - on Windows they must specify the target using environment variable, such as `set ARCH=arm` or `set ARCH=arm64`  , which will then be respected by the rest of logic that installs dependencies via `vcpkg`.

- if users don't specify, most likely they are building for Intel or ARM, nothing exotic.. And generic arch lookup would look like this (bulky!). This is much smarter than what I had done before, but doesn't really add much additional benefit.. 

```cmake

if(DEFINED ENV{ARCH})
  # Architecture may be specified via ARCH environment variable
  set(ARCH $ENV{ARCH})
else()
  # Autodetection logic that populates ARCH environment variable
  if(CMAKE_SYSTEM_PROCESSOR MATCHES "amd64.*|x86_64.*|AMD64.*")
    # Windows may report AMD64 even if target is 32-bit
    if(CMAKE_SIZEOF_VOID_P EQUAL 8)
      # 64 bits
      set(ARCH x64)
    elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
      # 32 bits
      set(ARCH x86)
    endif()
  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "i686.*|i386.*|x86.*")
    set(ARCH x86)
  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(aarch64.*|AARCH64.*|arm64.*|ARM64.*)")
    set(ARCH arm64)
  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(arm.*|ARM.*)")
    set(ARCH arm)
  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)64le")
    set(ARCH ppc64le)
  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)64")
    set(ARCH ppc64)
  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(mips.*|MIPS.*)")
    set(ARCH mips)
  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(riscv.*|RISCV.*)")
    set(ARCH riscv)
  else()
    message(FATAL_ERROR "opentelemetry-cpp: unrecognized target processor configuration!")
  endif()
endif()
message(STATUS "Building for architecture: ARCH=${ARCH}")

```

The only benefit is if we ever want to auto-install dependencies via `vcpkg` on Linux or Mac, for example, then knowing the target triplet comes in handy. It's a bit of an overkill, and I don't see a simpler solution. I looked at how Samsung did in in their repo, and for Windows they just assume that users manually specify the target arch by parameter.

lalitb reviewed (approved):
LGTM.

maxgolov commented:
@ThomsonTan - I added a bit smarter logic, that should reliably cover Intel/AMD-Windows and ALL-Linux/-Mac.

Re. ARM cross-compile only ... My assumption for `arm` and `arm64` cross-compiling on Windows:

- if developers compile using `tools/build.cmd` , then it respects the `ARCH` set via environment variable, which subsequently gets passed down to CMake. The script also accepts the `ARCH` as 1st parameter.

- if developers use Visual Studio 2019 IDE with CMake to cross-compile on Windows for ARM/ARM64, then they can add `ARCH=ARM64` variable to `CMakeSettings.json` using `environments` section. I also have an expectation that even if they don't, the new logic that I added MAY work with newer CMake, assuming that newer CMake is smarter with setting up the proper `CMAKE_SYSTEM_PROCESSOR`.

There is another even more smart way, to compiler a test program just to determine what arch we are compiling for.. But that'd be already too much. For what it's worth - passing an env var `ARCH` explicitly solves any possible ambiguity about what target ARCH we are compiling for.

lalitb commented on CMakeLists.txt:
```diff
@@ -9,6 +9,46 @@ project(opentelemetry-cpp)
+    endif()
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES
+         "^(aarch64.*|AARCH64.*|arm64.*|ARM64.*)")
+    set(ARCH arm64)
```
If we are adding this, should we also be checking for 32 bit OS running on 64bit ARM, and set ARCH accordingly?

maxgolov replied:
No. The issue with unreliable arch detection is unique to Windows (`MSVC`) only.

Linux and Mac are always Okay.

I am not aware of developers building our SDK on Windows-ARM64 host machine. I am not sure even if such MSVC toolset exists. If those developers using ARM64 host as their developer box exist somewhere, then they may need to explicitly specify the `ARCH` environment variable.

ThomsonTan replied:
Windows ARM64 have x86/x64 MSVC run under emulation. There is no native ARM64 MSVC as for now.

ThomsonTan commented on examples/grpc/CMakeLists.txt:
```diff
@@ -39,12 +39,7 @@ endif()
-    example_grpc_proto
-    protobuf::libprotobuf
-    gRPC::grpc++
-    gRPC::grpc++_reflection
```
Why is `grpc++_reflection` removed?

maxgolov replied:
Because there is no such package / target / library, at least not in `vcpkg` build of gRPC. Can you point me where this target is declared?

maxgolov replied:
I might be mistaken on this one. But it all compiles without this target. I see where it is mentioned in gRPC build. But I had an issue with that one on Windows.. we do build all examples across all CI loops for CMake?

ThomsonTan replied:
It is a library target as below. I think we build all examples in our CMake CI. We probably don't use use the reflection function in gRPC.

https://github.com/grpc/grpc/blob/master/CMakeLists.txt#L3146

lalitb replied:
Yes, all examples are built so hopefully should be fine :) Though I faintly remember getting link errors without this library, but all good if it builds fine in CI.
